Adding keyword to Reload the configuration of the KubeLibrary

Nilsty opened the pull request:
Please review this PR after https://github.com/devopsspiral/KubeLibrary/pull/7 has been merged.

In my tests for Humanitec, I need to be able to verify if a K8s job in one cluster has triggered a deployment for another cluster. For this, I need to change the connected Cluster during my test. To be able to do this I added a keyword that runs the `__init__` function of the library again with a different configuration. `Reload Config`

(Running the builtin keyword `Import Library` would create a second instance of the same Library causing all keywords to be duplicated.)

I have not added any tests for this keyword as I'm struggling to define a test setup for it. Maybe you have an idea @devopsspiral ?

m-wcislo commented:
That is great feature. I guess everything boils down to having more than one kubeconfig. I think we can get close enough with testing/examples with having two kubeconfigs with different context (different namespace) and just switching between them in example, you already created new namespace, we can assume other example/test prerequisites works on default namespace.

Also I guess we need to switch back at the end of test not to affect other tests in suite.

If somehow this mechanism becomes complicated it would be great to add some unit tests too to library itself.

m-wcislo commented on src/KubeLibrary/KubeLibrary.py:
```diff
@@ -58,6 +58,18 @@ def __init__(self, kube_config=None, incluster=False, cert_validation=True):
+        - ``cert_validation``:
+          Default True. Can be set to False for self-signed certificates.
+        """
+        self.__init__(kube_config=kube_config, incluster=incluster, cert_validation=cert_validation)
```
I think it is safer to have some refugular function here, instead of __init__. Like:
```
def __init__(...)
  reload_config()

def reload_config(...):
   <content of current init>
```

Nilsty replied:
ok, understood. I'll update it like this.

m-wcislo commented:
Also there is something like ConnectionCache in RF, this way we could switch between named sessions - just an idea I saw today, not sure if you need it here.

https://robot-framework.readthedocs.io/en/3.0/_modules/robot/utils/connectioncache.html
https://github.com/peterservice-rnd/robotframework-winrmlibrary/blob/2b318f304379d7c9f83b891b899fab89da5638d5/src/WinRMLibrary.py#L55

Nilsty commented:
Ok. I'll look into it. For now I found a good way to run a test for reloading the config via KinD. With KinD it is easy to have two different clusters locally.

Nilsty commented:
> Also there is something like ConnectionCache in RF, this way we could switch between named sessions - just an idea I saw today, not sure if you need it here.
> 
> https://robot-framework.readthedocs.io/en/3.0/_modules/robot/utils/connectioncache.html
> https://github.com/peterservice-rnd/robotframework-winrmlibrary/blob/2b318f304379d7c9f83b891b899fab89da5638d5/src/WinRMLibrary.py#L55

I looked into this. Some of the ConnectionCache functionality seems to be exposed with Builtin Keywords
[Get Library Instance](http://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Get%20Library%20Instance)
[Reload Library](http://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Reload%20Library)
[Set Library Search Order](http://robotframework.org/robotframework/latest/libraries/BuiltIn.html#Set%20Library%20Search%20Order)
I was able to have multiple instances of the KubeLibrary with different configurations using these keywords, but I was not able to make the right search order for those library instances work. Overall I believe it's cleaner and more readable in tests to provide a Reload functionality from the KubeLibrary itself.

Nilsty commented:
I think I'm happy with the PR now. Please review @devopsspiral 
Here is a list of overall changes:
Added Keyword **Reload Config** to change cluster connection within a test
Added Keyword **List Namespaces**
Updated the README, added KinD setup for testing with multiple clusters, restructuring example test case section
Updated the documentation with the new keywords
Added a test case example for **Reload Config** based on the KinD multi cluster setup

m-wcislo commented on README.md:
````diff
@@ -41,6 +41,7 @@ More examples in testcases/ directory.
 
+### Cluster Tests
 ```
 git clone https://github.com/kyma-incubator/octopus
````
octopus and grafana should be moved to respective sections. Cluster tests needs only bare k3s cluster.

m-wcislo commented on src/KubeLibrary/KubeLibrary.py:
```diff
@@ -77,6 +89,15 @@ def k8s_api_ping(self):
                                            _return_http_data_only=False)
         return resp
 
+    def list_namespaces(self):
```
We might name it get_namespaces to match the other ones.

m-wcislo commented on testcases/reload-config/reload-config_kw.robot:
```diff
@@ -0,0 +1,21 @@
+
+*** Keywords ***
+Connected to cluster-1 
+    Import Library  ${CURDIR}/../../src/KubeLibrary/KubeLibrary.py    kube_config=./cluster1-conf
```
It would be best to use KubeLibrary from package here i.e.:
 ```
Import Library    KubeLibrary    kube_config=./cluster1-conf
```

just to be consistent, even knowing it will not work currently.

Nilsty commented:
Thanks for your feedback. I updated the PR accordingly.
